File: solutions/dart/secret-handshake/1/lib/secret_handshake.dart
// Sacredly dedicated to Shree DR.MDD

extension on int {
  int fetchBit(int pos) => (this >> pos) & 1;
  bool toggleAt(int pos) => fetchBit(pos) == 1;
}

const SIGNALS = <String>['wink', 'double blink', 'close your eyes', 'jump'];

class SecretHandshake {
  List<String> commands(int encoded) {
    var output = <String>[];

    for (var idx = 0; idx < SIGNALS.length; idx++) {
      if (encoded.toggleAt(idx)) {
        output.add(SIGNALS[idx]);
      }
    }

    return encoded.toggleAt(SIGNALS.length) ? output.reversed.toList() : output;
  }
}

File: solutions/typescript/secret-handshake/1/secret-handshake.ts
// Dedicated to Shree DR.MDD
export function commands(code: number) {
  const binStr = numToBinary(code)
  const steps = binStr.split('').reverse().map((bit, idx) => {
    if (bit === '1') {
      return decodeGesture(idx)
    }
  }).filter(item => item)
  return binStr[0] === '1' ? steps.reverse() : steps
}

function decodeGesture(pos: number): string | undefined {
  const gestures = [
    'wink',
    'double blink',
    'close your eyes',
    'jump',
  ]
  if (pos >= 0 && pos < gestures.length) {
    return gestures[pos]
  }
}

function numToBinary(val: number): string {
  let bitPos = 4
  let binResult = ''
  while (bitPos >= 0) { 
    if (val >= (2 ** bitPos)) {
      val -= (2 ** bitPos)
      binResult += '1'
    } else {
      binResult += '0'
    }
    bitPos--
  }
  return binResult
}
